doc/source/guide/in_depth/functional_guide.rst

```diff
@@ -0,0 +1,257 @@
+
+        def __init__(self, domain, y):
+            # This comand calls the init of Functional and sets a number of
+            # parameters associated with a functional. All but domain have default
+            # values if not set.
+            super().__init__(domain=domain, linear=False, convex=True,
+                             concave=False, smooth=True, grad_lipschitz=2)
+
+            # We need to check that y is in the domain. Then we store the value of
+            # y for future use.
+            if y not in domain:
+                raise TypeError('y is not in the domain!')
+            self._y = y
+
+        # Property that returns the linear term.
+        @property
+        def y(self):
+            return self._y
+
+        # Defining the _call function
+        def _call(self, x):
+            return x.norm()**2 + x.inner(self.y)
+
+        # Next we define the gradient. Note that this is a property.
+        @property
+        def gradient(self):
+
+            # The class corresponding to the gradient operator.
+            class MyGradientOperator(odl.Operator):
+                """Class that implements the gradient operator of the functional
+                ``||x||_2^2 + <x,y>``.
+                """
+
+                def __init__(self, functional):
+                    super().__init__(domain=functional.domain,
+                                     range=functional.domain)
+
+                    self._functional = functional
+
+                def _call(self, x):
+                    return 2.0 * x + self._functional.y
+
+            return MyGradientOperator(functional=self)
+
+        # Next we define the convex conjugate functional.
+        @property
+        def conjugate_functional(self):
+            # This functional is implemented below.
+            return MyFunctionalConjugate(domain=self.domain, y=self.y)
+
+
+    # Here is the conjugate functional.
+    class MyFunctionalConjugate(odl.solvers.Functional):
+        """Conjugate functional to ``||x||_2^2 + <x,y>``.
+
+        Calculations give that this funtional has the analytic expression
+        f^*(x) = ||x||^2/2 - ||x-y||^2/4 + ||y||^2/2 - <x,y>.
+        """
+        def __init__(self, domain, y):
+            super().__init__(domain=domain, linear=False, convex=True,
+                             concave=False, smooth=True, grad_lipschitz=2)
+
+            if y not in domain:
+                raise TypeError('y is not in the domain!')
+            self._y = y
+
+        @property
+        def y(self):
+            return self._y
+
+        def _call(self, x):
+            return (x.norm()**2 / 2.0 - (x - self.y).norm()**2 / 4.0 +
+                    self.y.norm()**2 / 2.0 - x.inner(self.y))
+
+With this code, one can now create things like the conjugate functional of a scaled and translated version::
+
+    my_func = MyFunctional(domain=space, linear_term=linear_term)
+    (scalar * my_func).translate(translation).conjugate_functional
```

Comment:
In general a really good walk-through of all properties defined by `Functional`. Maybe a short section on "How to use the properties" or the likes would be interesting for users, too. For example,

- Using the steepest descent method with `my_functional.gradient`
- Using `my_functional.proximal` and `my_functional.convex_conj.proximal` in the "simplest" proximal method we have (not sure what that is)
- Using some method written for arbitrary operators to show that all that stuff still works